src/combinatorics/young/young_diagram.lean

```diff
@@ -305,4 +305,100 @@ end
 
 end row_lens
 
+section equiv_list_row_lens
+/-! ### Equivalence between Young diagrams and lists of natural numbers
+
+This section defines the equivalence between Young diagrams `mu` and weakly decreasing lists `w`
+of positive natural numbers, corresponding to row lengths of the diagram:
+  `young_diagram.equiv_list_row_lens :`
+  `young_diagram ≃ {w : list ℕ // w.sorted (≥) ∧ ∀ x ∈ w, 0 < x}`
+
+The two directions are `young_diagram.row_lens` (defined above) and `young_diagram.of_row_lens`.
+
+-/
+
+/-- The cells making up a `young_diagram` from a list of row lengths -/
+protected def cells_of_row_lens : list ℕ → finset (ℕ × ℕ)
+| [] := finset.empty
+| (w :: ws) := (({0} : finset ℕ) ×ˢ finset.range w) ∪
+                 (cells_of_row_lens ws).map
+                   (function.embedding.prod_map ⟨_, nat.succ_injective⟩ (by refl))
+
+protected lemma mem_cells_of_row_lens {w : list ℕ} {c : ℕ × ℕ} :
+  c ∈ young_diagram.cells_of_row_lens w ↔ ∃ (h : c.fst < w.length), c.snd < w.nth_le c.fst h :=
+begin
+  induction w generalizing c;
+  rw young_diagram.cells_of_row_lens,
+  { change false ↔ ∃ (h : _ < 0), _, simp },
+  { rcases c with ⟨⟨_, _⟩, _⟩,
+    { simp },
+    { simpa [w_ih, -finset.singleton_product, nat.succ_lt_succ_iff] } }
+end
+
+/-- Young diagram from a sorted list -/
+def of_row_lens (w : list ℕ) (hw : w.sorted (≥)) : young_diagram :=
+{ cells        := young_diagram.cells_of_row_lens w,
+  is_lower_set := begin
+    rintros ⟨i2, j2⟩ ⟨i1, j1⟩ ⟨hi : i1 ≤ i2, hj : j1 ≤ j2⟩ hcell,
+    rw [finset.mem_coe, young_diagram.mem_cells_of_row_lens] at hcell ⊢,
+    obtain ⟨h1, h2⟩ := hcell,
+    refine ⟨hi.trans_lt h1, _⟩,
+    calc j1 ≤ j2            : hj
+      ...   < w.nth_le i2 _ : h2
+      ...   ≤ w.nth_le i1 _ : _,
+    cases eq_or_lt_of_le hi,
+    { subst h },
```

Comment:
I think it's nice writing it like this because it makes the variables for the cases explicit, and it does the `subst` for you. It doesn't save having to close the first case though, so you still need `refl`.
```suggestion
    obtain (rfl | h) := eq_or_lt_of_le hi,
    { refl },
```